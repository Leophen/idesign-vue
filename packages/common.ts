import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
