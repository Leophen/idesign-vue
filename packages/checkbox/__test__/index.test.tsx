import { vi } from 'vitest';
import { Checkbox, CheckboxGroup } from '../index';
import { render } from '../../common'

describe('Checkbox 组件测试', () => {
  it('create', () => {
    const wrapper = render(<Checkbox />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('checked & children', () => {
    const wrapper = render(<Checkbox modelValue={true}>广州</Checkbox>)
    expect(wrapper.classes()).toContain('i-checkbox-is-checked');
    expect(wrapper.find('.i-checkbox__label').text()).toBe('广州');
  });

  it('defaultChecked', () => {
    const wrapper = render(<Checkbox defaultChecked={true} />)
    expect(wrapper.classes()).toContain('i-checkbox-is-checked');
  });

  it('disabled', () => {
    const fn = vi.fn();
    const wrapper = render(<Checkbox disabled={true} onChange={fn} />)
    expect(wrapper.classes()).toContain('i-checkbox-is-disabled');
    wrapper.find('.i-checkbox__former').trigger('change');
    expect(fn).toBeCalledTimes(0);
  });

  it('size', () => {
    const wrapper = render(<Checkbox size="small" />)
    expect(wrapper.classes()).toContain('i-checkbox--size-small');
  });
});

describe('CheckboxGroup 组件测试', () => {
  it('selected', () => {
    const wrapper = render(
      <CheckboxGroup modelValue={['sz']}>
        <Checkbox value="gz">广州</Checkbox>
        <Checkbox value="sz" data-testid='sz-item'>深圳</Checkbox>
      </CheckboxGroup>,
    );
    expect(wrapper.find('[data-testid="sz-item"]').classes()).toContain('i-checkbox-is-checked');
  });

  it('defaultSelected', () => {
    const wrapper = render(
      <CheckboxGroup defaultChecked={['gz']}>
        <Checkbox value="gz" data-testid='gz-item'>广州</Checkbox>
        <Checkbox value="sz">深圳</Checkbox>
      </CheckboxGroup>,
    );
    expect(wrapper.find('[data-testid="gz-item"]').classes()).toContain('i-checkbox-is-checked');
  });

  it('disabled', () => {
    const fn = vi.fn();
    const wrapper = render(
      <CheckboxGroup disabled>
        <Checkbox value="gz" data-testid='gz-item'>广州</Checkbox>
        <Checkbox value="sz">深圳</Checkbox>
      </CheckboxGroup>,
    );
    expect(wrapper.find('[data-testid="gz-item"]').classes()).toContain('i-checkbox-is-disabled');
    wrapper.find('[data-testid="gz-item"]').trigger('change');
    expect(fn).toBeCalledTimes(0);
  });

  it('size', () => {
    const wrapper = render(
      <CheckboxGroup size="small">
        <Checkbox value="gz" data-testid='gz-item'>广州</Checkbox>
        <Checkbox value="sz">深圳</Checkbox>
      </CheckboxGroup>,
    );
    expect(wrapper.find('[data-testid="gz-item"]').classes()).toContain('i-checkbox--size-small');
  });
});

