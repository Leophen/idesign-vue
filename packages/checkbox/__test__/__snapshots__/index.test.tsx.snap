// Vitest Snapshot v1

exports[`Checkbox ç»„ä»¶æµ‹è¯• > create 1`] = `
<label
  class="i-checkbox i-checkbox--size-medium"
>
  <input
    class="i-checkbox__former"
    readonly=""
    type="checkbox"
  />
  <span
    class="i-checkbox__input"
  />
  <span
    class="i-checkbox__label"
  >
    <!---->
  </span>
</label>
`;
